src/server/project.ts

```diff
@@ -2105,7 +2105,22 @@ namespace ts.server {
         }
 
         updateReferences(refs: readonly ProjectReference[] | undefined) {
-            this.projectReferences = refs;
+            // @ts-ignore
+            if (process.versions.pnp) {
+                const getPnpPath = (path: string) => {
+                    const basePath = this.getCurrentDirectory();
+                    const absolutePath = ts.getNormalizedAbsolutePath(path, basePath);
+                    try {
+                        return require("pnpapi").resolveToUnqualified(require(`${absolutePath}/package.json`).name, `${basePath}/`);

```

Comment:
I think it's best if we retrieve the package name from the PnP API, or at the very least don't use `require` - otherwise there are security aspects (for instance if a file named `package.json.js` exists, etc). The TS team is very against possible runtime executions, so better leave it at a minimum.

```suggestion
                    const pnpApi = require(`pnpapi`);
                    
                    const basePath = this.getCurrentDirectory();
                    const absolutePath = ts.getNormalizedAbsolutePath(path, basePath);

                    const targetLocator = pnpApi.findPackageLocator(absolutePath);
                    // eslint-disable-next-line no-null/no-null
                    if (targetLocator === null) {
                      return path;
                    }
                    
                    const locator = 
                    try {
                        return pnpApi.resolveToUnqualified(targetLocator.name, `${basePath}/`);

```